inspector: allow concurrent inspector sessions

eugeneo opened the pull request:
This change enables concurrent inspector sessions, through WebSocket
interface as well as JS interface, in any combination.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

jasnell commented:
Generally good with this, haven't reviewed the code yet so won't sign off just yet. But, I wanted to ask if you could articulate some of the issues that could present with this. For instance, suppose I have two concurrent sessions open... is it possible for one to interfere with the other? For instance, what impact does setting a breakpoint in one session have on the other? Can there be hidden side effects?

eugeneo commented:
@jasnell there are many edge cases, and that is expected...

The main use-case had been to allow different tools use different "domains" concurrently - e.g. some IDE would connect to Chrome for debugging, while Chrome DevTools can still be used to inspect DOM. For Node, this should allow debuginng the application that uses the `inspector` module.

1. Breakpoints do not get reported to other frontends, to avoid confusing them.
2. Pause events are reported to all frontends.
3. "Debugger.resume" sent by any frontend will resume execution for all.

eugeneo commented:
Another CI run, I had a bad testcase in the original version: https://ci.nodejs.org/job/node-test-pull-request/14367/

jasnell commented:
> 1. Breakpoints do not get reported to other frontends, to avoid confusing them.
> 2. Pause events are reported to all frontends.
> 3. "Debugger.resume" sent by any frontend will resume execution for all.

1 and 2 are the concerns for me. What we'll end up with are pauses in one frontend that happens for no obvious reason. What I'd like to make sure of is that when the pause is reported, there is enough context included for one frontend to know that the pause originated in a concurrently connected session so at least there's the ability to reason about it. If that's already accounted for, then awesome, never mind me :-)

eugeneo commented:
@jasnell this isn't that different from the `debugger` statement. Frontends already have to be ready for spurious pauses.

eugeneo commented:
Found a bug in V8 inspector around multisession support. Updated this PR with a workaround for a said bug. New CI: https://ci.nodejs.org/job/node-test-pull-request/14373/

jasnell commented:
@eugeneo ... thank you for the additional detail. As I've said I'm generally good with this I just want to make sure we have a good idea of what issues may come up with this. Any time we get into multi-session anything, especially with single loop, there are *bound* to be issues ;-) . Thank you for doing this!

eugeneo commented:
CI for a new version: https://ci.nodejs.org/job/node-test-pull-request/14386/

eugeneo commented:
CI mostly green, looks like MacOS CI is busted and failes on all changes. There are no inspector-specific test failures.

TimothyGu commented on src/inspector_agent.cc:
```diff
@@ -363,20 +365,22 @@ class NodeInspectorClient : public V8InspectorClient {
-        new ChannelImpl(client_.get(), delegate));
+  int connectFrontend(std::unique_ptr<InspectorSessionDelegate> delegate) {
+    events_dispatched_ = true;
+    int session_id = next_session_id_++;
```
Is it possible for this to have a race condition (e.g. WebSocket server and a JS inspector session launching at the same time)?

eugeneo replied:
All API in this file is called only from the main thread. E.g. incoming WS session will "interrupt" V8 (or will use uv_async_t) before calling this function.

TimothyGu commented on src/inspector_agent.cc:
```diff
@@ -554,20 +564,25 @@ void Agent::Stop() {
+    // There is a bug in inspector that causes every session
+    // call V8InspectorClient::quitMessageLoopOnPause. We are using this flag
+    // to ignore those calls so message loop is spinning as long as there's a
+    // reason to expect inspector messages
```
Is there a crbug for this?

eugeneo replied:
https://crbug.com/834056

TimothyGu replied:
Can you put that link in the comment?

eugeneo replied:
I added it in a separate commit, will squash before merging.

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -363,20 +365,22 @@ class NodeInspectorClient : public V8InspectorClient {
+    events_dispatched_ = true;
+    int session_id = next_session_id_++;
+    channels_[session_id] =
+        std::make_unique<ChannelImpl>(client_.get(), std::move(delegate));
```
`client_.get()` looks kind of questionable.  Could it be turned into a `std::shared_ptr`?

eugeneo replied:
I do not want to imply shared ownership. I changed the code to use const std::unique_ptr&.

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -464,20 +464,31 @@ class NodeInspectorClient : public V8InspectorClient {
+  std::unordered_map<int, std::unique_ptr<ChannelImpl>> channels_;
   std::unordered_map<void*, InspectorTimerHandle> timers_;
+  int next_session_id_;
+  bool events_dispatched_;
```
`bool events_dispatched_ = false;`?

eugeneo replied:
Changed this and other similar fields.

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -554,20 +564,25 @@ void Agent::Stop() {
+std::unique_ptr<InspectorSession> Agent::Connect(
+    std::unique_ptr<InspectorSessionDelegate> delegate) {
+  int session_id = client_->connectFrontend(std::move(delegate));
+  return std::make_unique<InspectorSession>(session_id, client_.get());
```
Maybe turn `client_` into a `std::shared_ptr`?

eugeneo replied:
Did that.

bnoordhuis commented on src/inspector_js_api.cc:
```diff
@@ -125,18 +101,22 @@ class JSBindingsConnection : public AsyncWrap {
-    session->CheckIsCurrent();
-    Agent* inspector = env->inspector_agent();
-    inspector->Dispatch(ToProtocolString(env->isolate(), info[0])->string());
+    if (session->session_)
```
Can you put braces around the body?

eugeneo replied:
Done

bnoordhuis commented on src/inspector_js_api.cc:
```diff
@@ -125,18 +101,22 @@ class JSBindingsConnection : public AsyncWrap {
 
+static bool InspectorEnabled(Environment* env) {
+  Agent* agent = env->inspector_agent();
+  return !!agent->io() || agent->HasConnectedSessions();
```
Can you do an explicit nullptr check instead of coercing to bool?

eugeneo replied:
Done

bnoordhuis commented on test/parallel/test-inspector-multisession-ws.js:
```diff
@@ -0,0 +1,75 @@
+}, 150);
+session.on('Debugger.paused', () => {
+  session.disconnect();
+  process.exit(0);
```
It would be infinitesimally nicer to have node exit naturally by calling `clearInterval(interval)`.

eugeneo replied:
Done

eugeneo commented:
Thank you for the reviews, I addressed the comments.
New CI: https://ci.nodejs.org/job/node-test-pull-request/14405/

I will merge this PR after the weekend.

Trott commented:
Relevant macOS test failure:

https://ci.nodejs.org/job/node-test-commit-osx/17976/nodes=osx1010/console

```console
not ok 1107 parallel/test-inspector-multisession-ws
  ---
  duration_ms: 0.330
  severity: fail
  exitcode: 1
  stack: |-
    [test] Connecting to a child Node process
    [test] Testing /json/list
    [test] Connecting to a child Node process
    [test] Testing /json/list
    [err] Debugger listening on ws://127.0.0.1:50554/d51f55a4-567c-49b0-89cd-669ce131d100
    [err] For help, see: https://nodejs.org/en/docs/inspector
    [err] 
    { Error: write EPIPE
        at WriteWrap.afterWrite [as oncomplete] (net.js:836:14) errno: 'EPIPE', code: 'EPIPE', syscall: 'write' }
    1
  ...
```

eugeneo commented:
Test failure is caused by https://github.com/nodejs/node/pull/20254, I would like to land that PR before landing this one.

ofrobots commented:
Given this from statement @Trott:
> Blocking on the test failure. Feel free to clear this if the test has been fixed.

Since the CI is green now, I think it is reasonable to dismiss that review.

eugeneo commented:
Landed as https://github.com/nodejs/node/commit/13001035340d6d3fb173fef4a1db42ba70f61720

addaleax commented:
Just got this CI failure for an AIX stress-test run: https://ci.nodejs.org/job/node-stress-single-test/1825/nodes=aix61-ppc64/console

It seems related to this PR? Not sure why it wouldn’t show up in regular CI…

eugeneo commented:
Should I remove `make_unique` or this is old failure and new AIX now uses newer STL?

richardlau commented:
@eugeneo I believe it's an infrastructure issue with the stress job not using the correct level of C++ compiler on AIX (the normal CI jobs should be fine).

This PR is the first change that's actually caused the build to fail on older compilers (where previously you would be warned but the build would succeed), but there's a reason we specify the [supported toolchains](https://github.com/nodejs/node/blob/master/BUILDING.md#supported-toolchains).

eugeneo commented:
@richardlau Ok, thanks. Ping me if this needs to be fixed - this is a trivial change.
